gax/operation.h

```diff
@@ -0,0 +1,118 @@
+// Copyright 2019 Google Inc.  All rights reserved
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GAPIC_GENERATOR_CPP_GAX_OPERATION_H_
+#define GAPIC_GENERATOR_CPP_GAX_OPERATION_H_
+
+#include "call_context.h"
+#include "internal/gtest_prod.h"
+#include "operations_stub.h"
+#include "status.h"
+#include "status_or.h"
+
+#include "google/longrunning/operations.pb.h"
+
+#include <iostream>
+
+#include <memory>
+
+namespace google {
+
+namespace gax {
+
+// Note: compilation will fail unless ResponseT and MetadataT are children of
```

Comment:
You can enforce that with `std::enable_if<std::is_base_of<protobuf::Message, ResponseT>::value>`, e.g.:

```C++
template <typename ResponseT, typename MetadataT,
typename std::enable_if<std::is_base_of<protobuf::Message, ResponseT>::value, int>::type = 0,
typename std::enable_if<std::is_base_of<protobuf::Message, MetadataT>::value, int>::type = 0>
class Operation final {
```

The main advantage is that you get an error earlier, and often more readable, than just letting the compiler figure out that you are missing something.

Reply:
Good to know. I'd been reading about concepts and how they had been punted to C++20 and wasn't sure there was a good mechanism for doing the same thing in 11.